File: gen_traffic.c
#define M 100000         // Maximum number of demands

#define N 12            	// Number of Nodes. 
#define L 34		// Number of directed Links.
#define S 50		// Number of slots per link 	400

#define A 500		// Traffic load
#define H 10		// 1/mu average holding time
#define req_Max 6	// Maximum demand size			16

#define K 100

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <fstream>
#include <iostream>
#include <iomanip>
#include <random>

using namespace std;

int main(void)
{
	int t_req[M], t_hold[M], t_exp[M];
	
	int hops[N][N];
	int link[N][N];

	int max_hold;
	
	unsigned seed1 = 123;
	unsigned seed2 =  125;

	int lp_size[M], source[M], dest[M];
	
	int i,j;
	int arr_int;
	double temp1;
	
	double mu = 1/double(H);  // for exponentially-distributed duration with mean H=1/mu
	double inter_arr = double(H)/A;  // Inter-arrival time for poisson distribution A=H/inter_arr
	
	srand (seed2);	// Use srand = time(NULL) to generate data that change every time
	
	default_random_engine generator (seed1);
	poisson_distribution<int> next_arr(K*inter_arr);  // 100* to clock the steps to a 10ms, does not change A, 
	exponential_distribution<double> hold_time(mu);	  // A = 100*H/100*inter_arriv 
	uniform_int_distribution<int> traff_dist(1, req_Max);
	
	ofstream ofs1;
    	ofs1.open("traffic_jpn12_.txt");
	if(!ofs1){
		cout<< "Cannot open input_demands1 file"<<endl;
		return 1;
	}
	
	//ofs1 << "mu and inter:=" << mu <<", "<<  inter_arr<< endl;
	//ofs1 << "LP number, s, d, size, Arrival time, holding time:=" << endl;
	
	max_hold = 0;
	t_req[0]= 0;
	
	for (int i=0; i<M; ++i){
		temp1 = hold_time(generator);
		t_hold[i] = int(K*temp1)+1;							//100*
		if(max_hold < t_hold[i]) max_hold = t_hold[i];
		arr_int = next_arr(generator);
		lp_size[i] = traff_dist(generator);
		source[i] = rand() %N;
		dest[i] = rand() %N;
			while(source[i] == dest[i]) dest[i] = rand() %N;
		t_exp[i] = t_req[i] + t_hold[i];
		ofs1 << i  << ": "<< source[i]  <<" " << dest[i]  <<" " << lp_size[i]  <<" " << t_req[i] <<" " << t_hold[i] << endl;
		t_req[i+1]= t_req[i] + arr_int;
	}
	
	//ofs1 << ":;"<< endl << endl;
	ofs1.close();
}

File: gen_traffic.py
import random
import numpy as np

# Constants
M = 100000  # Maximum number of demands
N = 12  # Number of Nodes
L = 34  # Number of directed Links
S = 50  # Number of slots per link
A = 500  # Traffic load
H = 10  # 1/mu average holding time
req_Max = 6  # Maximum demand size
K = 100


def main():
    t_req = [0] * (M + 1)  # +1 for t_req[i+1] access in the loop
    t_hold = [0] * M
    t_exp = [0] * M

    hops = [[0 for _ in range(N)] for _ in range(N)]
    link = [[0 for _ in range(N)] for _ in range(N)]

    max_hold = 0

    seed1 = 123
    seed2 = 125

    slot_indx = [0] * M
    lp_size = [0] * M
    source = [0] * M
    dest = [0] * M

    # Random generators setup
    mu = 1.0 / H  # for exponentially-distributed duration with mean H=1/mu
    inter_arr = H / A  # Inter-arrival time for poisson distribution A=H/inter_arr

    random.seed(seed2)
    np.random.seed(seed1)

    # Open output file
    try:
        with open("traffic_jpn12_100_updated.txt", "w") as ofs1:
            # Initialize first arrival time
            t_req[0] = 0

            for i in range(M):
                # Generate holding time (exponential distribution)
                temp1 = np.random.exponential(1 / mu)
                t_hold[i] = int(K * temp1) + 1

                if max_hold < t_hold[i]:
                    max_hold = t_hold[i]

                # Generate arrival interval (Poisson)
                arr_int = np.random.poisson(K * inter_arr)

                # Generate traffic demand size
                lp_size[i] = random.randint(1, req_Max)

                # Generate source and destination
                source[i] = random.randint(0, N - 1)
                dest[i] = random.randint(0, N - 1)
                
                slot_indx[i] = random.randint(0, S - lp_size[i])

                '''
                Nếu source[j] == source[i] và slot_indx[j] in range (slot_indx[j], slot_indx[j] + lp_size[j]):
                    thì generate lại slot_indx[j]
                Hoặc không để cho nó giống nhau
                ''' 
                # Ensure source and destination are different
                # while source[i] == dest[i]:
                #     dest[i] = random.randint(0, N - 1)

                # Calculate expiration time
                t_exp[i] = t_req[i] + t_hold[i]

                # Write to file
                ofs1.write(f"{i}: {source[i]} {dest[i]} {lp_size[i]} {slot_indx[i]} {t_req[i]} {t_hold[i]}\n")

                # Calculate next arrival time
                t_req[i + 1] = t_req[i] + arr_int

    except IOError:
        print("Cannot open input_demands1 file")
        return 1


if __name__ == "__main__":
    main()